code_tools/2 type_checkers/index.md

```diff
@@ -0,0 +1,570 @@
+
+The `typing` module provides more duck types, most notably: `Sequence` and `Mapping`. `Sequence` is a duck type for anything that keeps an order and is index-able. Lists and tuples are `Sequence`s, but a `set` for instance is not.
+
+
+    from typing import Sequence
+
+    a: Sequence[int] = [1, 2, 3]  # All good
+    b: Sequence[int] = (1, 2, 3)  # All good
+    c: Sequence[int] = {1, 2, 3}  # Incompatible types in assignment (expression has type "Set[int]", variable has type "Sequence[int]")
+
+
+`Mapping` is a generic type for structures that map one value to another, such as dictionaries for instance.
+
+
+    from typing import Mapping
+
+    a: Mapping[str, int] = {"foo": 1}  # All good
+
+
+<details markdown="1"><summary  markdown="span">For the technically curious...</summary>
+
+These abstract data types are implemented as so called `Protocols`. See this [Python Enhancement Proposal](https://www.python.org/dev/peps/pep-0544/). Through these Protocols you can define your own duck types too. For instance:
+
+
+    from typing import Iterable, Protocol
+
+    class SupportsAdd(Protocol):
+        def __add__(self, other):
+            pass
+
+    def sum(items: Iterable[SupportsAdd]) -> SupportsAdd:
+        total = None
+        for item in items:
+            if total is None:
+                total = item
+            else:
+                total += item
+        return item
+
+    sum([1, 2, 3]) # all good
+    sum([1.5, None]) # error: List item 1 has incompatible type "None"; expected "SupportsAdd"
+
+
+</details>
+
+</details>
+
+4. Annotate the code below with duck types instead:
+
+
+        T = TypeVar("T")
+
+        def reverse(items: list[T]) -> list[T]:
+            new = []
+            for item in items:
+                new.insert(0, item)
+            return new
+
+
+    <textarea name="form[q4.1]" rows="5" required=""></textarea>
+
+
+        T = TypeVar("T")
+
+        def select(items: list[T], indices: list[int]) -> list[T]:
+            selection = []
+            for index in indices:
+                selection.append(items[index])
+            return selection
+
+    
+    <textarea name="form[q4.2]" rows="5" required=""></textarea>
+
+
+
+        T = TypeVar("T")
+
+        def filter(items: list[T], allowed: dict[T, bool]) -> list[T]:
+            new = []
+            for item in items:
+                if allowed[item]:
+                    new.append(item)
+            return new
+
+
+    <textarea name="form[q4.3]" rows="6" required=""></textarea>
+
+<details markdown="1"><summary  markdown="span">Nominal subtyping aka subclassing </summary>
```

Comment:
Twijfel of dit nog interessant is voor het niveau van de studenten.

Reply:
Ik ook. Ik plaatste dit vooral voor heuristieken, waar ze vooral met hun eigen classes gaan werken.

Reply:
Is het dan optioneel? Of hoe zie je dat voor je?

Reply:
Voor de volledigheid, staan ook geen oefeningen bij

Reply:
OK maar wat is nu de rol van zo'n stukje tekst. Moeten ze het lezen, begrijpen, leren?

Reply:
Lezen en begrijpen. Naar mijn idee is het niet al te ingewikkeld.